jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DownloadObjectAction.kt

```diff
@@ -34,28 +41,114 @@ class DownloadObjectAction(private val project: Project, treeTable: S3TreeTable)
-        val destination = FileChooser.chooseFile(descriptor, project, baseDir)?.path ?: return
-        downloadAll(project, files.map { it.key to Paths.get(destination, it.name).outputStream() })
+        val destination = getDownloadLocation(true) ?: return
+
+        val downloads = files.map { DownloadInfo(it.key, destination.resolve(it.name)) }
+        val finalDownloads = checkForConflicts(destination, downloads)
+
+        downloadAll(project, finalDownloads)
     }
 
-    private fun downloadSingle(project: Project, file: S3TreeObjectNode) {
+    private fun getDownloadLocation(multipleFiles: Boolean): Path? {
         val baseDir = VfsUtil.getUserHomeDir()
-        val descriptor = FileSaverDescriptor(message("s3.download.object.action"), message("s3.download.object.description"))
-        val destination = FileChooserFactory.getInstance().createSaveFileDialog(descriptor, project).save(baseDir, file.name)?.file?.outputStream() ?: return
-        downloadAll(project, listOf(file.key to destination))
+
+        val descriptor = if (multipleFiles) {
+            FileChooserDescriptorFactory.createSingleFolderDescriptor().also {
+                it.description = message("s3.download.object.browse.description.multiple")
+            }
+        } else {
+            FileChooserDescriptorFactory.createSingleFileOrFolderDescriptor().also {
+                it.description = message("s3.download.object.browse.description.single")
+            }
+        }
+
+        descriptor.title = message("s3.download.object.browse.title")
+
+        // In order to prevent a confusing UX around when we can have folders and not, force the non native chooser
+        descriptor.isForcedToUseIdeaFileChooser = true
+
+        return FileChooser.chooseFile(descriptor, project, baseDir)?.path?.let {
+            Paths.get(it)
+        }
     }
 
-    private fun downloadAll(project: Project, files: List<Pair<String, OutputStream>>) {
+    private fun checkForConflicts(targetDirectory: Path, files: List<DownloadInfo>): List<DownloadInfo> {
+        val finalDownloads = mutableListOf<DownloadInfo>()
+
+        files.forEachIndexed { index, download ->
+            if (!download.diskLocation.exists()) {
+                finalDownloads.add(download)
+                return@forEachIndexed
+            }
+
+            val description = message(
+                "s3.download.object.conflict.description",
+                targetDirectory.relativize(download.diskLocation),
+                targetDirectory
+            )
+
+            val options = if (files.size == 1) {
+                arrayOf(
+                    message("s3.download.object.conflict.skip"),
+                    message("s3.download.object.conflict.overwrite")
+                )
+            } else {
+                arrayOf(
+                    message("s3.download.object.conflict.skip"),
+                    message("s3.download.object.conflict.overwrite"),
```

Comment:
IMO we should build this second array out of the first just so that there is no chance they are misordered